### PR TITLE
Add test plan for group functions

### PR DESCRIPTION
**Pull request by Melirius**
Adds a test plan for group functions as described in sections 4.17.3 "Group functions" and 4.17.4 "Group algorithms library" of the SYCL 2020 specification.

### CODE REVIEW

File: test_plans/group_functions.asciidoc
Comment (Melirius) [reply test_plans/group_functions.asciidoc:85]: Definitely tests should be run for group, I add a sentence about it also. BTW, am I right that group algorithms library functions (4.17.4) can be run also on subgroup?

File: test_plans/group_functions.asciidoc
Comment (Melirius) [reply test_plans/group_functions.asciidoc:124]: No, we test that function exist and also that it accepts aforementioned memory scopes. I'll change the wording, thanks.

It is really difficult to meaningfully test barrier but minimum positive testing can be done. I'll add it.

File: test_plans/group_functions.asciidoc
Comment (Melirius) [reply test_plans/group_functions.asciidoc:130]: Yep, I mean it when said about different `Predicate`,  but it should be expressed explicitly.

File: test_plans/group_functions.asciidoc
Comment (gmlueck) [reply test_plans/group_functions.asciidoc:86]: > Definitely tests should be run for group, I add a sentence about it also. BTW, am I right that group algorithms library functions (4.17.4) can be run also on subgroup?

Yes.  All of those functions have a statement like:

> Available only if `sycl::is_group_v<std::decay_t<Group>>` is true

And we say in [4.17.3.1](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_group_type_trait):

> `is_group<T>` is `std::true_type` if `T` is the type of a standard SYCL group (`group` or `sub_group`)

File: test_plans/group_functions.asciidoc
Comment (gmlueck) [inline test_plans/group_functions.asciidoc:128]: I think you will need to write the test so that it checks that all these memory scopes are supported by using `info::device::atomic_fence_scope_capabilities`.  The specification says that only `memory_scope::work_group` is required to be supported.  The test should skip any memory scopes that are not supported according to that query.

Adding @pennycook here ... should the spec also guarantee that `memory_scope::sub_group` is supported?  The default memory  scope for `group_barrier<sub_group>()` is `memory_scope::sub_group`.  It seems weird to have a default value that might not be supported by the device.  In any case, isn't it trivial to support `memory_scope::sub_group` if you are required to support `memory_scope::work_group`?

Comment (Melirius) [reply test_plans/group_functions.asciidoc:129]: Yes, you're right, it should be tested before.

File: test_plans/group_functions.asciidoc
Comment (gmlueck) [inline test_plans/group_functions.asciidoc:124]:
```suggestion
If supported by the hardware (from query of `info::device::atomic_fence_scope_capabilities`)
```

**Review by keryell — APPROVED** [review]
Looks  good.
@Pennycook by looking at all this list in a central place, I was wondering whether the naming is right in the specification for the presence of `group` or not and the place of `group` in the various functions or member functions.

**Comment by fraggamuffin** [comment]
fix then can merge

File: test_plans/group_functions.asciidoc
Comment (bader) [reply test_plans/group_functions.asciidoc:129]: @gmlueck, @Pennycook, should I wait for @bashbaug to reply before merging this PR?

Comment (bashbaug) [reply test_plans/group_functions.asciidoc:129]: Note: I think this is one of the places where SYCL and OpenCL can be different, if desired, though we'll need some care if they are different.

In OpenCL in general, the "subgroup" memory scope is only required when subgroups are supported.

* In OpenCL C it only exists when subgroups are supported ([link](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#memory-scope) - "Requires support for OpenCL C 3.0 or newer and the __opencl_c_subgroups feature.").
* In SPIR-V, the subgroup memory scope is only valid when subgroups are supported ([link](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Env.html#validation-rules) - "For OpenCL 2.1, OpenCL 2.2, or OpenCL 3.0 devices supporting Subgroups (where CL_DEVICE_MAX_NUM_SUB_GROUPS is not 0)."  See note below...).
* This means that a device that does not support subgroups does not need to handle a subgroup scope memory fence, say by "upgrading" it to the wider work-group scope.

If SYCL chooses to require the subgroup scope when the work-group scope is also supported it's fine to do so, but it will be SYCL's responsibility to generate valid SPIR-V in this case, specifically by generating a work-group fence instead of a subgroup fence if/when subgroup fences aren't supported.

I agree with most of @pennycook's suggested clarifications above, especially that we should require that the subgroup scope is supported when subgroups are supported.  This is debatably already true, but it is not stated explicitly.

When we do this I also think it makes sense to describe the SPIR-V requirements in terms of the atomic fence capabilities query, rather than subgroups generically.  Again, this is debatably already true, but I think it's a bit clearer and consistent to describe the requirements in this way.

Comment (gmlueck) [reply test_plans/group_functions.asciidoc:129]: I think we can merge this test plan before this conversation is resolved.  The only impact on the test plan is whether the test can assume that `memory_scope::sub_group` is supported or if it must check to see if the device supports this memory scope.

I suggest that the CTS follow the existing wording of the SYCL spec, which does *not* guarantee that `memory_scope::sub_group` is supported.

I have opened the following issue to continue the conversation about whether the SYCL spec should be changed:

https://gitlab.khronos.org/sycl/Specification/-/issues/639

**Pull request merged**
